### PR TITLE
Multiple applications support (Phase 1)

### PR DESCRIPTION
**Pull request by noomorph**
- [x] This change has been discussed in issue #2563 and the solution has been agreed upon with maintainers.

---

**Description:**

- [x] Add new definitions for the config
- [x] Implement new device config resolution
- [x] Implement new apps config resolution
- [x] Implement device.selectApp(name) method
- [x] In `composeAppsConfig.js`: reimplement `cliConfig.appLaunchArgs` override (temporary deleted at the moment)
- [x] In `composeAppsConfig.test.js`: cover unhappy scenarios for: colliding apps, property name typos, mismatches between app and device platform, restore launch args tests, etc...
- [x] implement multiple `detox build` capability and improve some warnings
- [x] In `detox init` – adapt template to the new configuration schema
- [x] Move `utilBinaryPaths` to device configuration
- [x] Expand `DetoxConfigErrorBuilder.test.js` to cover all new methods
- [x] In `DetoxConfigErrorBuilder.js`: replace stub methods with real validation error creation:
- [x] Forcibly relaunch app on `device.launchApp()` after selecting another app — overall, the best workaround at the moment is to terminate the current app when selecting another one. While theoretically it is not required, practically this is the safest choice considering that we have a single session id for all the apps. 😞 We might improve that later.
- [x] Add a small e2e test for switching between the apps
- [x] Update the docs

### CODE REVIEW

File: detox/src/artifacts/screenshot/ScreenshotArtifactPlugin.js
Comment (d4vidi) [inline detox/src/artifacts/screenshot/ScreenshotArtifactPlugin.js:20]: Hmm... okay?

File: detox/test/detox.config.js
Comment (d4vidi) [inline detox/test/detox.config.js:2]: Shouldn't we remove this from `package.json`, in that case?

**Review by d4vidi — APPROVED** [review]
See some minor comments 🚀

File: detox/src/artifacts/screenshot/ScreenshotArtifactPlugin.js
Comment (noomorph) [reply detox/src/artifacts/screenshot/ScreenshotArtifactPlugin.js:20]: Totally :) there is no argument there at all. Noticed thanks to TypeScript, actually.

File: detox/test/detox.config.js
Comment (noomorph) [reply detox/test/detox.config.js:2]: Yep, forgot. 😛

**Review by d4vidi — APPROVED** [review]

**Comment by d4vidi** [comment]
@noomorph, approved, but note the main test suite is broken on CI.

**Comment by d4vidi** [comment]
Heads up, closing and reopening to verify a subtle issue with jenkins. Not related to this PR in any way.

File: detox/index.d.ts
Comment (d4vidi) [inline detox/index.d.ts:172]:
```suggestion
        interface DetoxAndroidAppConfig extends DetoxLooseAndroidAppConfig {
```

File: detox/index.d.ts
Comment (d4vidi) [inline detox/index.d.ts:176]: I understand where the loose concept is coming from but since they are never directly referenced - only the concrete types do, then maybe it'd be best to simplify by unifying everything onto the concrete types and that's it?

Or alternatively, create one "loose" config that includes 'em properties that are relevant to both (i.e. all) platforms (e.g `launchArgs`)? And then reconsider the name loose, and try to go with conventions such as "abstract" or just "base"?... 🤔

File: detox/index.d.ts
Comment (d4vidi) [inline detox/index.d.ts:201]: (nitpicking-alert) move this lower to reside with the actual config flavors?

File: detox/index.d.ts
Comment (d4vidi) [inline detox/index.d.ts:364]: I'm kind-of picking my brain trying to think whether this is really needed, because none of the API's are really app-specific (or are they?)

File: detox/index.d.ts
Comment (noomorph) [reply detox/index.d.ts:176]: Mmm, I can see your point - let me check one more time.

File: detox/index.d.ts
Comment (noomorph) [reply detox/index.d.ts:364]: This is a protected method - I use it internally with multiple apps config to unselect the default app so you always will be getting an error unless you explicitly select your app after Detox init.

Do you think it is better to remove it at all? I don't have a strong opinion here, so pls let me know.

Comment (d4vidi) [reply detox/index.d.ts:364]: Oh, it _is_ `@protected` indeed :)
Your call, then!

But to be sure I understand correctly: eventually, this is so as to protect users from forgetting to select an app when multiple apps have been configured in the selected configuration?

Comment (noomorph) [reply detox/index.d.ts:364]: @d4vidi, exactly for that purpose — we can't decide what application should be default there.

File: detox/local-cli/build.js
Comment (d4vidi) [inline detox/local-cli/build.js:49]: This alone yields 💯^2 points in dev ux 🔥

File: detox/src/utils/environment.test.js
Comment (d4vidi) [inline detox/src/utils/environment.test.js:271]: ❤️

File: detox/test/detox.config.js
Comment (d4vidi) [inline detox/test/detox.config.js:38]: So clean... 'love it!

File: detox/test/detox.config.js
Comment (d4vidi) [inline detox/test/detox.config.js:40]: can it be just `ios`, maybe? (and `android` instead of `android.apk`?)

File: detox/test/detox.config.js
Comment (d4vidi) [inline detox/test/detox.config.js:168]: 'this here on purpose, in order to test config overrides?

File: detox/src/devices/LaunchArgs.js
Comment (d4vidi) [inline detox/src/devices/LaunchArgs.js:17]: and also... why? 😢

File: detox/src/devices/LaunchArgs.js
Comment (d4vidi) [inline detox/src/devices/LaunchArgs.js:11]: but... but why?

File: detox/src/devices/Device.js
Comment (d4vidi) [inline detox/src/devices/Device.js:81]: (Rubber-duck'ing) So this will work but would omit any previous changes made in the context of the selected app. For example:
```js
device.selectApp('app1');
device.appLaunchArgs.modify(...);
device.selectApp('app2'); // changes to app1's args are gone...
device.selectApp('app1'); // ... so here we start from scratch
```

That's ok under the assumption of that `selectApp()` is meant to be called once per test. Was that the (implicit) assumption here?

File: detox/src/devices/Device.js
Comment (d4vidi) [inline detox/src/devices/Device.js:96]: Sign this deprecation up for Detox 19 😄

**Review by d4vidi — COMMENTED** [review]
Still a WIP, but this is what I have so far. Good work, overall! I hope to be able to share more notes soon.

File: detox/src/configuration/index.test.js
Comment (d4vidi) [inline detox/src/configuration/index.test.js:39]: As a general note (not directly related to the PR), this test is a bit too integrative to my taste. With the various configuration domain each broken down to its own composer (session, runner, apps, etc) along with the awesome unit testing coverage -- I feel a total mocking of each with some minimal glue-logic testing here should suffice.

I know the approach is arguable, no need to accept it. In any case, no need to address this right now.

File: detox/src/devices/Device.js
Comment (d4vidi) [inline detox/src/devices/Device.js:46]: (as discussed 1x1) Consider revisiting the implicitness of this approach.
Maybe as a tradeoff, a mere rename (e.g `_currentAppArgs`) might help make it more understandable.
Or to solve altogether - what if a LaunchArgs object is created per each app, to replace the plain config? Just a suggestion.

**Review by d4vidi — COMMENTED** [review]
Some more notes after going over the changes in Device and the various composers.
I couldn't go very deep with those I'm afraid. I trust that real-life feedback would be more beneficial, anyways!

File: detox/src/devices/Device.js
Comment (noomorph) [reply detox/src/devices/Device.js:81]: > but... but why?
> and also... why? 😢

Exactly to avoid this scenario you demonstrated.

File: detox/src/devices/Device.js
Comment (noomorph) [reply detox/src/devices/Device.js:96]: AFAIR, we spoke with Leo about this — and looks like we might refrain from deprecating it at all.

File: detox/test/detox.config.js
Comment (noomorph) [reply detox/test/detox.config.js:168]: Yeah — but maybe it is too weird to keep it this way.

File: detox/src/Detox.js
Comment (noomorph) [reply detox/src/Detox.js]: Fixed according to your comments - see https://github.com/wix/Detox/pull/2587/commits/4ccc073a2302d281ddeb1788d5ef7f8281137b89

File: detox/src/Detox.test.js
Comment (noomorph) [reply detox/src/Detox.test.js:168]: I've deleted `installUtilBinaries` assertion there, but not sure what did you mean exactly?

File: detox/src/Detox.js
Comment (noomorph) [reply detox/src/Detox.js:200]: thanks. 🤘

File: detox/src/utils/environment.test.js
Comment (noomorph) [reply detox/src/utils/environment.test.js:271]: Better works for Windows as well.

File: detox/test/detox.config.js
Comment (noomorph) [reply detox/test/detox.config.js:38]: Pleased to ~hear~ read 😄

File: detox/src/devices/LaunchArgs.js
Comment (noomorph) [reply detox/src/devices/LaunchArgs.js:17]: answered below. looks like it is going to be a temporary workaround

File: detox/src/devices/LaunchArgs.js
Comment (noomorph) [reply detox/src/devices/LaunchArgs.js:11]: answered below. looks like it is going to be a temporary workaround

File: detox/src/devices/Device.js
Comment (noomorph) [reply detox/src/devices/Device.js:81]: Anyway, we've discussed the motivation in 1:1 mode — although, good to replace that mutation-reliant approach with some more sturdy.

File: detox/src/devices/Device.js
Comment (noomorph) [reply detox/src/devices/Device.js:46]: I'll add a TODO after the merge

File: detox/src/configuration/index.test.js
Comment (noomorph) [reply detox/src/configuration/index.test.js:39]: Let's add a TODO to discuss this later.

**Pull request merged**
